### DISCUSSION TITLE
Snapshot concept explanation: how is the snapshot ID per run calculated

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#8553] [Q&A]
Hello all. Could anyone talk (or give a link to the subject) about the concept of snapshots ("Snapshot ID" column in Dagit). In particular how is the ID per Run (Launch) calculated? Does it make sense to get a snapshot object (or it is just internal dagster entity)? If so how can I use and get the snapshot ID for runs with the same ID with API (graphql?) without direct queries to db tables? I can't find the subject in the docs :disappointed:. Thanks in advance.

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1638340129217200?thread_ts=1638340129.217200&cid=C01U954MEER).*

**Comment by yuhan** [comment]
The snapshot ID column refers to a "snapshot" of the structure of the job/pipeline at the time the run occurred. This is useful in the case that you've changed your job's structure in some way, the old run will still be able to refer back to the snapshot of the way your job used to look, and will be interpretable that way.

To be clear, we don't store the actual job itself, so you can't rerun the job using this snapshot / old structure. It's more for historical / metadata uses. There's most likely a way to retrieve these via graphql, but it's definitely more of an internal concept. The job snapshot ID itself is a hash of the serialized structure of the job - only meaningful changes to the structure should change that hash.

However, because we don't expect users to be directly interacting with these snapshots, it's possible that across releases the snapshot structure itself may change, so even when user code has not changed at all, the snapshot ID might change as a result of that.
